### ISSUE TITLE
App behavior is different with PublishTrimmed=true (PublishSingleFile=true) on linux

### ISSUE THREAD

**Issue by NitzanEgozy** [mono/linker#864]
(This is also posted on https://github.com/dotnet/sdk/issues/4042, as I'm not entirely sure where is the right place for this to be. I'll close the irrelevant one when I understand better)

When packing with both PublishTrimmed=true and PublishSingleFile=true, I see that the application behaves different. I've attached a sample of ILSpy Console application (ICSharp.Code.Console, originally from https://github.com/icsharpcode/ILSpy). This is a slightly modified version of 5.0.1 - removed all non-used project files and changed sdk to 3.1. I've also attached an input .exe files for the application which has different outputs if the application is published with/without PublishTrimmed=true.
Note that it only happens when I publish it on linux - windows works perfect.

You can find the replication here:
https://github.com/NitzanEgozy/dotnet/tree/master/testers/dotnet_trimmed_bug
1. cd dotnet_trimmed_bug
2. run.sh
3. Compare the outputs of ILSpyInconsistentExe/decompile_output_trimmed and ILSpyInconsistentExe/decompile_output_non_trimmed (more specifically - look at the decompiled DynamicClass.cs - you'll see many errors in the decompilation after trimming)

For example, look at this method:
non-trimmed
```
public override bool TrySetMember(SetMemberBinder binder, object value)
		{
			if (_fields.ContainsKey(binder.Name))
			{
				Type type = _fields[binder.Name].Key;
				if (value.GetType() == type)
				{
					_fields[binder.Name] = new KeyValuePair<Type, object>(type, value);
					return true;
				}
				throw new Exception("Value " + value?.ToString() + " is not of type " + type.Name);
			}
			return false;
		}
```
trimmed
```	
public override bool TrySetMember(SetMemberBinder binder, object value)
		{
			//IL_0023: Unknown result type (might be due to invalid IL or missing references)
			//IL_0028: Unknown result type (might be due to invalid IL or missing references)
			//IL_0050: Unknown result type (might be due to invalid IL or missing references)
			if (_fields.ContainsKey(binder.get_Name()))
			{
				global::System.Type type = _fields.get_Item(binder.get_Name()).get_Key();
				if (value.GetType() == type)
				{
					_fields.set_Item(binder.get_Name(), new KeyValuePair<global::System.Type, object>(type, value));
					return true;
				}
				throw new global::System.Exception("Value " + value?.ToString() + " is not of type " + ((MemberInfo)type).get_Name());
			}
			return false;
		}
```
(Also - this is just a small replication. On more complicated logic this gives unreadable decompilation results)
I don't think the output itself gives any good information - this is merely a hint that something in the packaging has gone wrong (some missing reference maybe?)

**Comment by MichalStrehovsky** [comment]
I've looked at ILSpy from trimming perspective in the past, so I'll jot down my notes: could be unrelated, but someone investigating this might find it helpful:

1. ILSpy uses the Microsoft.VisualStudio.Composition MEF engine.
2. As part of its operation ILSpy/MEF scans for all DLL files in the directory looking for interesting things in them.
3. It then uses Reflection.Emit to generate code that does things with the found code.

Removing unused code in an app like this is a task similar to removing unused code in JavaScript (it's so dynamic that static tools don't stand a chance). I wouldn't be surprised if even the very conservative setting of IL Linker used in .NET Core ends up removing something that was necessary after all (especially for point 2 above).

**Comment by NitzanEgozy** [comment]
Thanks for the response!
I will just add that I do not look at ILSpy executable - I'm looking at the console application, which only linked to netstandard dlls (as far as I could find).
I am trying to dig deep into finding which references are missing, so any note is helpful!

**Comment by NitzanEgozy** [comment]
So, this seems as though it is a problem with ILSpy. Not much of a problem, but the design of ILSpy uses it's own framework files for decompilation resolving. The following code is the problematic area:
`	public class UniversalAssemblyResolver : IAssemblyResolver
	{
...
...
		string ResolveInternal(IAssemblyReference name)
		{
...
			var framework_dir = Path.GetDirectoryName(typeof(object).Module.FullyQualifiedName);
...
			assembly = SearchDirectory(name, framework_dirs);
			if (assembly != null)
				return assembly;
...
`
Basically, when he sees that the target dll references libraries, it checks to see if they exist where the current executable is running from. For self-contained application, this is `var/tmp/.net/ilspycmd/[randomly-generated]`. When you trim, it's just missing some dlls that the target dll is referencing - which are not used by ILSpy, so they were trimmed.
This leads me to believe that ILSpy shouldn't be trimmed - or require tons of work to make it less-reliant on local framework dlls
